File: ios/SovereignCommunications/Core/JSBridge.swift
//
//  JSBridge.swift
//  SovereignCommunications
//
//  Bridge to @sc/core Javascript runtime.
//

import Foundation
import JavaScriptCore
import os.log

@objc protocol JSBridgeExports: JSExport {
    func onOutboundMessage(_ peerId: String, _ data: String) // Base64 encoded string for data
    func onApplicationMessage(_ messageJson: String)
}

class JSBridge: NSObject, JSBridgeExports {
    static let shared = JSBridge()
    private let logger = Logger(subsystem: "com.sovereign.communications", category: "JSBridge")
    
    private var context: JSContext?
    private var coreNetwork: JSValue?
    
    var outboundCallback: ((String, Data) -> Void)?
    var applicationMessageCallback: ((String) -> Void)? // JSON string
    
    private override init() {
        super.init()
        setupContext()
    }
    
    private func setupContext() {
        context = JSContext()
        
        context?.exceptionHandler = { context, exception in
            if let exception = exception {
                self.logger.error("JS Exception: \(exception.toString() ?? "unknown")")
            }
        }
        
        // Inject self to handle callbacks from JS
        context?.setObject(self, forKeyedSubscript: "NativeBridge" as NSString)
        
        // Load the Core Bundle
        if let bundlePath = Bundle.main.path(forResource: "sc-core", ofType: "js") {
            do {
                let script = try String(contentsOfFile: bundlePath)
                context?.evaluateScript(script)
                
                // Initialize MeshNetwork
                // Assumes Global 'SCCore' object exposes MeshNetwork
                let initScript = """
                const network = new SCCore.MeshNetwork();
                
                // Register outbound transport
                network.registerOutboundTransport(async (peerId, data) => {
                    // Convert Uint8Array to Base64 to send across bridge
                    const base64 = SCCore.utils.bytesToBase64(data); // Assuming utility exists or we use Buffer
                    NativeBridge.onOutboundMessage(peerId, base64);
                });
                
                // Register application message listener
                network.onMessage((msg) => {
                    NativeBridge.onApplicationMessage(JSON.stringify(msg));
                });
                
                globalThis.meshNetwork = network;
                """
                context?.evaluateScript(initScript)
                
                coreNetwork = context?.objectForKeyedSubscript("meshNetwork")
                logger.info("JS Core initialized successfully")
            } catch {
                logger.error("Failed to load sc-core.js: \(error.localizedDescription)")
            }
        } else {
            logger.error("sc-core.js not found in bundle")
        }
    }
    
    // MARK: - Native Inputs
    
    /**
     * Pass incoming binary data from Native Transport (BLE/WebRTC) to JS Core
     */
    func handleIncomingPacket(data: Data, from peerId: String) {
        guard let network = coreNetwork else { return }
        
        // Convert Data to Uint8Array via Base64 or direct array
        // Easy way: Pass Base64 -> JS converts to Uint8Array
        let base64 = data.base64EncodedString()
        
        let script = """
        const data = SCCore.utils.base64ToBytes("\(base64)");
        meshNetwork.handleIncomingPacket("\(peerId)", data);
        """
        context?.evaluateScript(script)
    }
    
    /**
     * Send message from UI (Application Layer) -> JS Core -> Mesh
     */
    func sendMessage(to recipientId: String, content: String) {
        let script = """
        meshNetwork.sendTextMessage("\(recipientId)", "\(content)");
        """
        context?.evaluateScript(script)
    }
    
    // MARK: - JSBridgeExports
    
    func onOutboundMessage(_ peerId: String, _ dataBase64: String) {
        if let data = Data(base64Encoded: dataBase64) {
            outboundCallback?(peerId, data)
        }
    }
    
    func onApplicationMessage(_ messageJson: String) {
        applicationMessageCallback?(messageJson)
    }
}

File: android/app/src/main/kotlin/com/sovereign/communications/service/JSBridge.kt
package com.sovereign.communications.service

import android.content.Context
import android.util.Log
import com.eclipsesource.v8.V8
import com.eclipsesource.v8.V8Object
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

/**
 * Bridge to @sc/core Javascript runtime.
 * Hosting the V8 engine (via J2V8).
 */
class JSBridge(
    private val context: Context,
) {
    private val TAG = "JSBridge"
    private val scope = CoroutineScope(Dispatchers.Main)

    private var v8: V8? = null

    // Callback for outbound messages: (peerId, data) -> Unit
    var outboundCallback: ((String, ByteArray) -> Unit)? = null

    // Callback for application messages: (jsonString) -> Unit
    var applicationMessageCallback: ((String) -> Unit)? = null

    init {
        setupContext()
    }

    private fun setupContext() {
        Log.d(TAG, "Initializing JS Context...")
        try {
            v8 = V8.createV8Runtime()

            // Register NativeBridge callbacks
            val nativeBridge = V8Object(v8)
            v8?.add("NativeBridge", nativeBridge)

            nativeBridge.registerJavaMethod({ _, parameters ->
                val peerId = parameters.getString(0)
                val dataBase64 = parameters.getString(1)
                onOutboundMessage(peerId, dataBase64)
            }, "onOutboundMessage")

            nativeBridge.registerJavaMethod({ _, parameters ->
                val messageJson = parameters.getString(0)
                onApplicationMessage(messageJson)
            }, "onApplicationMessage")

            // Clean up V8Object handle
            nativeBridge.close()

            // Load sc-core.js from assets
            val jsContent =
                context.assets
                    .open("sc-core.bundle.js")
                    .bufferedReader()
                    .use { it.readText() }
            v8?.executeVoidScript(jsContent)

            // Initialize MeshNetwork
            val initScript = """
                const network = new SCCore.MeshNetwork();
                network.registerOutboundTransport((peerId, data) => {
                    // Bridge back to Java
                    const base64 = SCCore.utils.bytesToBase64(data);
                    NativeBridge.onOutboundMessage(peerId, base64);
                });
                network.onMessage((msg) => {
                    NativeBridge.onApplicationMessage(JSON.stringify(msg));
                });
                globalThis.meshNetwork = network;
            """
            v8?.executeVoidScript(initScript)

            Log.i(TAG, "JS Context setup complete")
        } catch (e: Exception) {
            Log.e(TAG, "Failed to initialize V8", e)
        }
    }

    /**
     * Pass incoming binary data from Native Transport (BLE/WebRTC) to JS Core
     */
    fun handleIncomingPacket(
        data: ByteArray,
        from: String,
    ) {
        try {
            val base64 = android.util.Base64.encodeToString(data, android.util.Base64.NO_WRAP)
            // Use safe script execution
            val script = "if (globalThis.meshNetwork) { meshNetwork.handleIncomingPacket('$from', SCCore.utils.base64ToBytes('$base64')); }"
            v8?.executeVoidScript(script)
            Log.v(TAG, "Passed incoming packet from $from to JS Core (${data.size} bytes)")
        } catch (e: Exception) {
            Log.e(TAG, "Error handling incoming packet", e)
        }
    }

    /**
     * Send message from UI (Application Layer) -> JS Core -> Mesh
     */
    fun sendMessage(
        recipientId: String,
        content: String,
    ) {
        try {
            // Escape content for JS string
            val safeContent = content.replace("'", "\\'").replace("\n", "\\n")
            val script = "if (globalThis.meshNetwork) { meshNetwork.sendTextMessage('$recipientId', '$safeContent'); }"
            v8?.executeVoidScript(script)
            Log.d(TAG, "Sent message to $recipientId via JS Core")
        } catch (e: Exception) {
            Log.e(TAG, "Error sending message", e)
        }
    }

    // Java Interface methods called by JS
    fun onOutboundMessage(
        peerId: String,
        dataBase64: String,
    ) {
        val data = android.util.Base64.decode(dataBase64, android.util.Base64.NO_WRAP)
        scope.launch {
            outboundCallback?.invoke(peerId, data)
        }
    }

    fun onApplicationMessage(messageJson: String) {
        scope.launch {
            applicationMessageCallback?.invoke(messageJson)
        }
    }

    fun close() {
        try {
            v8?.close()
        } catch (e: Exception) {
            Log.e(TAG, "Error closing V8", e)
        }
    }
}
